feat(dataset): create a dataset using the ui

vfried opened the pull request:
This PR addresses #524 

To create the form I implemented a form generator that will allow us to create a form from from a Schema. This can then be reused in case we need to create other forms.

In this case:


**Schema:**
![image](https://user-images.githubusercontent.com/42647877/68942581-4105fb80-07a9-11ea-9b8e-5ce6f66bbb26.png)


**Call to the Form Generator (Form Panel)**
![image](https://user-images.githubusercontent.com/42647877/68942381-c6d57700-07a8-11ea-824c-a89bdc3df2b5.png)


**Generated Form**
![image](https://user-images.githubusercontent.com/42647877/68942061-d0121400-07a7-11ea-9a59-f326ed7f672b.png)

**Form with errors**
![image](https://user-images.githubusercontent.com/42647877/68942091-dd2f0300-07a7-11ea-9d33-f1c6d9a8c578.png)


The text editor uses CKEditor
https://ckeditor.com/docs/ckeditor5/latest/builds/guides/integration/frameworks/react.html

The file uploader uses FilePond
https://pqina.nl/filepond/

**### TO BE DONE:**
- Tests for the form generator
- Call to the backend to create the dataset and upload files (submitCallback)

ciyer commented:
The generic form framework is very cool!

vfried commented:
TO-DO:
- Improve upload of large files
- Progress bar for file uploads
- Fix redirection (a dataset is not visible until after you refresh the page), this could be because the KG is adding the dataset
- Retry button for when upload fails

ciyer commented on src/project/datasets/new/DatasetNew.present.js:
```diff
@@ -0,0 +1,105 @@
+    return <Col sm={12} md={8} lg={10}>
+    <Alert timeout={0} color="primary">
+      Acces Denied. You don't have rights to create datasets for this project.<br /><br /> 
+      <FontAwesomeIcon icon={faInfoCircle} />  If you where recently added to this project try refreshing the page.
```
Text: `If you were recently given access to this project, you might need to refresh the page first`.

Could we add a button that the user can click to refresh the page?

ciyer commented on src/utils/formgenerator/fields/FileUploaderInput.js:
```diff
@@ -0,0 +1,242 @@
+          return [];
+        } else {
+          let new_file = body.result.files[0];
+          if(body.result.files[0].file_name !== undefined && body.result.files[0].file_name !== file.name){
```
I think this can be:
```if (new_file.file_name !== undefined && new_file.file_name !== file.name) { ```

ciyer commented on src/utils/formgenerator/fields/FileUploaderInput.js:
```diff
@@ -0,0 +1,242 @@
+        } else {
+          let new_file = body.result.files[0];
+          if(body.result.files[0].file_name !== undefined && body.result.files[0].file_name !== file.name){
+            new_file = getFileObject(file.name, new_file.file_size, new_file.file_id, undefined, new_file.file_name)
```
Maybe call this variable `newFileObj` to avoid confusion

ciyer commented on src/utils/formgenerator/fields/FileUploaderInput.js:
```diff
@@ -0,0 +1,242 @@
+      droppedFiles.map(file => uploadFile(file));
+      setFiles([...files, ...droppedFiles]);
+      const newDisplayFiles = droppedFiles
+        .map(file => ( getFileObject(file.name, file.size, null, undefined, undefined)))
```
I think you don't need one set of parents:
```.map(file => getFileObject(file.name, file.size, null, undefined, undefined))```

ciyer commented on src/utils/formgenerator/fields/FileUploaderInput.js:
```diff
@@ -0,0 +1,242 @@
+                <td>{index + 1}</td>
+                <td>
+                  <span>{file.file_name}</span>
+                  {file.file_alias ? <small><br></br><span className="text-danger"> *File was renamed to:{file.file_alias}</span></small> : null}
```
We should probably mention why the file was renames:
```File name, <i>{file.file_name}</i>, contains disallowed characters; it has been renamed to <i>file.file_alias}</I>```

vfried commented:
I just pushed a fix for the requested changes, I also changed what happens after the form is submitted.

The function now:
- Waits until the dataset is created
- After that a timer is set where the function fetches the datasets of the project and checks if the dataset is there, if it is there the user is redirected the user to HTTPS://...datasets/new-dataset-id (Before this the timeout is cleared)

vfried commented:
Changed the text editor so it doesn't use the CKEditor and added message in case the dataset isn't found and was just created.

ciyer commented on src/project/datasets/DatasetsListView.js:
```diff
@@ -20,6 +21,23 @@ function DatasetListRow(props){
+        </DropdownToggle>
+        <DropdownMenu right={false}>
+          <DropdownItem>
+            <Link to={props.newDatasetUrl}>New Dataset</Link>
```
Could you also add a link to the newDatasetUrl to the top "+" menu?

 ![image](https://user-images.githubusercontent.com/1196411/73057084-e23bbd00-3e90-11ea-9007-8380a8852e9d.png)

vfried replied:
Done

ciyer commented on src/model/RenkuModels.js:
```diff
@@ -218,5 +219,45 @@ const notebooksSchema = new Schema({
+    name: 'name',
+    label: 'Name',
+    type: FormGenerator.FieldTypes.TEXT,
+    parseFun: expression => FormGenerator.Parsers.slugFromTitle(expression),
```
I like how you set the form up to automatically convert input into a slug. Very slick!

vfried commented:
I added a retry button for when the upload fails.

Added help text for the description of the dataset.

Regarding the 404 on the dataset, i do wait to see if the dataset was created in the KG and then redirect the user, i think that even though the kg knows it has a new dataset sometimes it can't fetch it. We should address this in a new pr though. I did add a message so the users know that the dataset could be there even if they get a 404.

Testing available in: datasetstest.dev.renku.ch

ciyer commented:
This is a very nice new feature!
